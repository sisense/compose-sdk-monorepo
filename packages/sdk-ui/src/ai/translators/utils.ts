export function toKebabCase(str: string): string {
  return str
    .replace(/\s+/g, '-') // Replace whitespace with hyphens
    .replace(/([a-z])([A-Z])/g, '$1-$2') // Convert camelCase to kebab-case
    .toLowerCase(); // Convert to lowercase
}

export function capitalizeFirstLetter(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}
