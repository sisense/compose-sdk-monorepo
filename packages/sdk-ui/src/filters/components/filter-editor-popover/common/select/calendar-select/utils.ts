import { DateLevels } from '@sisense/sdk-data';

import { SelectorMode } from '@/filters/components/date-filter/date-filter/calendar-date-selector';
import { getDefaultDateMask } from '@/query/date-formats/apply-date-format';

import { DatetimeFormatter } from '../../../hooks/use-datetime-formatter';
import { CalendarSelectTypes } from './types';

const MAX_CALENDAR_DISPLAY_ITEMS = 1;

export function getCalendarSelectedItemsDisplayValue(
  values: Date[],
  formatter: DatetimeFormatter,
): string | undefined {
  if (values.length === 0) {
    return undefined;
  }

  if (values.length > MAX_CALENDAR_DISPLAY_ITEMS) {
    return `${values.length} Dates selected`;
  }

  return values
    .map((dateValue) => formatter(dateValue, getDefaultDateMask(DateLevels.Days)))
    .join(', ');
}

export function getCalendarDateSelectorMode(type: CalendarSelectTypes): SelectorMode {
  switch (type) {
    case CalendarSelectTypes.SINGLE_SELECT:
      return 'pointSelector';
    case CalendarSelectTypes.MULTI_SELECT:
      return 'multiPointsSelector';
    case CalendarSelectTypes.RANGE_FROM_SELECT:
      return 'fromSelector';
    case CalendarSelectTypes.RANGE_TO_SELECT:
      return 'toSelector';
  }
}
