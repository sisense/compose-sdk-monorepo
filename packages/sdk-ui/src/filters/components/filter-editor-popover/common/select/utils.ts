import { PopoverAnchorPosition } from '@/common/components/popover';
import { SelectItem } from './types';

export const calculatePopoverPosition = (
  anchorEl: HTMLDivElement | null,
  contentHeight?: number,
): PopoverAnchorPosition => {
  const positionBottom: PopoverAnchorPosition = {
    anchorEl: anchorEl as HTMLElement,
    anchorOrigin: { vertical: 'bottom', horizontal: 'left' },
    contentOrigin: {
      vertical: 'top',
      horizontal: 'left',
    },
  };
  const positionTop: PopoverAnchorPosition = {
    anchorEl: anchorEl as HTMLElement,
    anchorOrigin: { vertical: 'top', horizontal: 'left' },
    contentOrigin: {
      vertical: 'bottom',
      horizontal: 'left',
    },
  };

  if (!anchorEl || !contentHeight) {
    return positionBottom;
  }

  const shouldPlaceOnBottom =
    window.innerHeight - anchorEl.getBoundingClientRect().bottom > contentHeight;

  return shouldPlaceOnBottom ? positionBottom : positionTop;
};

const MAX_DISPLAY_ITEMS = 3;

export function getSelectedItemsDisplayValue<Value>(
  items: SelectItem<Value>[],
  values: Value[],
): string | undefined {
  if (values.length === 0) {
    return undefined;
  }

  if (values.length > MAX_DISPLAY_ITEMS) {
    return `${values.length} selected`;
  }

  return items
    .filter((item) => values.includes(item.value))
    .map(({ value, displayValue }) => displayValue ?? value)
    .join(', ');
}
