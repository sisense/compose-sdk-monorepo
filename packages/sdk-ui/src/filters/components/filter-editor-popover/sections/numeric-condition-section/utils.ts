import { TFunction } from '@sisense/sdk-common';
import { Filter } from '@sisense/sdk-data';
import {
  filterToDefaultValues,
  filterToOption,
} from '../../../criteria-filter-tile/criteria-filter-operations.js';
import {
  isConditionalFilter,
  isExcludeMembersFilter,
  isNumericBetweenFilter,
  isSupportedByFilterEditor,
} from '../../utils.js';
import { isNumericString } from '@/utils/is-numeric-string.js';
import { NumericCondition, NumericConditionType, NumericConditionFilterData } from './types.js';
import {
  createExcludeMembersFilter,
  getConfigWithUpdatedDeactivated,
  getCriteriaFilterBuilder,
  getMembersWithoutDeactivated,
} from '../utils.js';

export function validateInputValue(value: string, t: TFunction) {
  if (!isNumericString(value)) {
    return t('filterEditor.validationErrors.invalidNumber');
  }

  return undefined;
}

export function getNumericFilterCondition(
  filter: Filter,
  initialCondition: NumericConditionType,
): NumericConditionType {
  if (
    !isConditionalFilter(filter) ||
    isNumericBetweenFilter(filter) ||
    !isSupportedByFilterEditor(filter)
  ) {
    // returns first condition by default
    return initialCondition;
  }

  if (isExcludeMembersFilter(filter)) {
    return NumericCondition.EXCLUDE;
  }

  return filterToOption(filter) as NumericConditionType;
}

export function getNumericFilterValue(filter: Filter): string {
  if (!isConditionalFilter(filter)) {
    return '';
  }
  const value = filterToDefaultValues(filter)[0] ?? '';
  return `${value}`;
}

export function createConditionalFilter(baseFilter: Filter, data: NumericConditionFilterData) {
  const { attribute } = baseFilter;
  const { condition, value, selectedMembers, multiSelectEnabled } = data;
  if (condition === NumericCondition.EXCLUDE) {
    const config = getConfigWithUpdatedDeactivated(baseFilter, selectedMembers);
    const members = getMembersWithoutDeactivated(baseFilter, selectedMembers);

    return createExcludeMembersFilter(attribute, members, {
      ...config,
      enableMultiSelection: multiSelectEnabled,
    });
  }

  if (!isNumericString(value)) {
    return null;
  }

  const builder = getCriteriaFilterBuilder(condition);
  return builder.fn(attribute, parseFloat(value), baseFilter.config);
}
