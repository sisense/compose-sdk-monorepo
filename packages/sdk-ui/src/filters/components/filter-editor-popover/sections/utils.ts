import {
  Attribute,
  DimensionalLevelAttribute,
  Filter,
  filterFactory,
  isMembersFilter,
  MembersFilterConfig,
} from '@sisense/sdk-data';
import { CRITERIA_FILTER_MAP } from '../../criteria-filter-tile/criteria-filter-operations.js';

export function getCriteriaFilterBuilder(condition: keyof typeof CRITERIA_FILTER_MAP) {
  return CRITERIA_FILTER_MAP[condition];
}

export function createExcludeMembersFilter(
  attribute: Attribute,
  members: string[],
  config?: MembersFilterConfig,
) {
  return members.length || config?.deactivatedMembers?.length
    ? filterFactory.members(attribute, members, {
        ...config,
        excludeMembers: true,
      })
    : null;
}

export function getMembersWithoutDeactivated(filter: Filter, selectedMembers: string[]) {
  return isMembersFilter(filter) && filter?.config?.deactivatedMembers
    ? selectedMembers.filter(
        (member: string) => !filter.config?.deactivatedMembers.includes(member),
      )
    : selectedMembers;
}
export function getMembersWithDeactivated(filter: Filter) {
  return isMembersFilter(filter) ? [...filter.members, ...filter.config.deactivatedMembers] : [];
}

export function getConfigWithUpdatedDeactivated(filter: Filter, selectedMembers: string[]) {
  return isMembersFilter(filter) && filter?.config?.deactivatedMembers
    ? {
        ...filter.config,
        deactivatedMembers: filter.config?.deactivatedMembers?.filter((member: string) =>
          selectedMembers.includes(member),
        ),
      }
    : filter.config;
}

/**
 * Returns the granularities that are restricted by the parent filters (previous cascading levels)
 *
 * @param datetimeAttribute - The datetime attribute
 * @param parentFilters - The parent filters
 * @returns The restricted granularities
 */
export function getRestrictedGranularities(
  datetimeAttribute: Attribute,
  parentFilters: Filter[] = [],
) {
  return parentFilters
    .filter((filter) => filter.attribute.expression === datetimeAttribute.expression)
    .map((f) => (f.attribute as DimensionalLevelAttribute).granularity);
}
