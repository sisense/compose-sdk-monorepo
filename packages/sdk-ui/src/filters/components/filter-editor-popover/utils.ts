import {
  Filter,
  isMembersFilter,
  MembersFilter,
  isMeasureFilter,
  isNumericFilter,
  isRankingFilter,
  isTextFilter,
  NumericOperators,
  isRelativeDateFilter,
  RelativeDateFilter,
  isCustomFilter,
} from '@sisense/sdk-data';

export function isSupportedByFilterEditor(filter: Filter): boolean {
  return !isRankingFilter(filter) && !isCustomFilter(filter);
}

export function isIncludeAllFilter(filter: Filter): filter is MembersFilter {
  return (
    isMembersFilter(filter) && !filter.members.length && !filter.config.deactivatedMembers.length
  );
}

export function isIncludeMembersFilter(filter: Filter): filter is MembersFilter {
  return isMembersFilter(filter) && !filter.config.excludeMembers;
}

export function isExcludeMembersFilter(filter: Filter): filter is MembersFilter {
  return isMembersFilter(filter) && filter.config.excludeMembers;
}

export function isConditionalFilter(filter: Filter) {
  return (
    isExcludeMembersFilter(filter) ||
    isMeasureFilter(filter) ||
    isNumericFilter(filter) ||
    isTextFilter(filter) ||
    isRankingFilter(filter)
  );
}

export function isNumericBetweenFilter(filter: Filter): boolean {
  return (
    isNumericFilter(filter) &&
    filter.operatorA === NumericOperators.From &&
    filter.operatorB === NumericOperators.To
  );
}

export function isRelativeDateFilterWithAnchor(filter: Filter): filter is RelativeDateFilter {
  return isRelativeDateFilter(filter) && !!filter.anchor;
}

export function isRelativeDateFilterWithoutAnchor(filter: Filter): filter is RelativeDateFilter {
  return isRelativeDateFilter(filter) && !filter.anchor;
}

/**
 * Formats a given Date object into a string in the format "YYYY-MM-DDT00:00:00".
 *
 * @param {Date} date - The Date object to format.
 * @returns {string} The formatted date string in "YYYY-MM-DDT00:00:00" format.
 */
export function convertDateToMemberString(date: Date): string {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');

  return `${year}-${month}-${day}T00:00:00`;
}
