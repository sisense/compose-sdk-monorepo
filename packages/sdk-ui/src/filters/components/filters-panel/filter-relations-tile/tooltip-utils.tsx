import {
  FilterRelationsDescription,
  isOpenBracketDescriptionNode,
  isCloseBracketDescriptionNode,
  isAttributeDescriptionNode,
  isOperatorDescriptionNode,
} from '@/utils/filter-relations';
import { TFunction } from '@sisense/sdk-common';
import { ReactNode } from 'react';

export function generateTooltipLines(
  filterRelationsDescription: FilterRelationsDescription,
  t: TFunction,
): ReactNode[] {
  const tooltipLines: ReactNode[] = [];
  let currentLine: ReactNode[] = [];
  const trimmedFilterRelationsDescription = trimUnnecessaryBrackets(filterRelationsDescription);
  trimmedFilterRelationsDescription.forEach((node) => {
    if (isOpenBracketDescriptionNode(node)) {
      currentLine.push(<>{'('}</>);
    } else if (isCloseBracketDescriptionNode(node)) {
      currentLine.push(<>{')'}</>);
    } else if (isAttributeDescriptionNode(node)) {
      currentLine.push(<i>{node.attribute}</i>);
    } else if (isOperatorDescriptionNode(node)) {
      // When encountering 'AND' or 'OR', push the current line and start a new one.
      tooltipLines.push(<>{...currentLine}</>);
      tooltipLines.push(
        <b>
          {node.operator === 'OR' ? <b>{t('filterRelations.or')}</b> : t('filterRelations.and')}
        </b>,
      );
      currentLine = [];
    }
  });
  tooltipLines.push(<>{...currentLine}</>);
  return tooltipLines;
}

/**
 * Recursively removes unnecessary "outermost" parentheses from the filter relations description.
 *
 * "Unnecessary" means a matching pair of open/close brackets that fully wraps
 * the entire expression, with no leftover operators or attributes outside.
 */
export function trimUnnecessaryBrackets(
  filterRelationsDescription: FilterRelationsDescription,
): FilterRelationsDescription {
  // If there's fewer than 2 nodes, it can't be wrapped in parentheses
  if (filterRelationsDescription.length < 2) {
    return filterRelationsDescription;
  }

  // Destructure the array into first node, middle slice, and last node
  const [first, ...rest] = filterRelationsDescription;
  const last = rest[rest.length - 1];
  const middle = rest.slice(0, -1);

  // If the first node isn't an open bracket or the last isn't a close bracket,
  // there's no "outermost" parentheses to remove
  if (!isOpenBracketDescriptionNode(first) || !isCloseBracketDescriptionNode(last)) {
    return filterRelationsDescription;
  }

  // Check if those outer parentheses fully wrap the expression inside.
  // We'll track bracket depth across the `middle` section using reduce.
  const { depth, isBalanced } = middle.reduce(
    (acc, node) => {
      // Increment or decrement depth when we see open or close brackets
      if (isOpenBracketDescriptionNode(node)) {
        return { ...acc, depth: acc.depth + 1 };
      }
      if (isCloseBracketDescriptionNode(node)) {
        const newDepth = acc.depth - 1;
        // If newDepth goes negative, it means the brackets aren't simply
        // wrapped at the outer level
        return {
          depth: newDepth,
          isBalanced: acc.isBalanced && newDepth >= 0,
        };
      }
      // For other node types, depth doesnâ€™t change
      return acc;
    },
    { depth: 0, isBalanced: true },
  );

  // If the entire middle is properly balanced (never negative, ends in zero depth),
  // the outer brackets are truly "wrapping" the entire expression and can be removed.
  if (isBalanced && depth === 0) {
    return trimUnnecessaryBrackets(middle);
  }

  // Otherwise, it's not purely wrapped, so we leave it as is
  return filterRelationsDescription;
}
