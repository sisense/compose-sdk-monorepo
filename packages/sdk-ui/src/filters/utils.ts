import { type Filter } from '@sisense/sdk-data';
import cloneDeep from 'lodash/cloneDeep';

/**
 * Clones a filter with a toggled 'disabled' state.
 * @param filter - Filter to clone
 */
export const cloneFilterAndToggleDisabled = <TFilter extends Filter>(filter: TFilter): TFilter => {
  const newFilter = cloneFilter(filter);
  newFilter.disabled = !filter.disabled;
  return newFilter;
};

export const cloneFilter = <TFilter extends Filter>(filter: TFilter): TFilter => {
  const newFilter = cloneDeep(filter);
  Object.setPrototypeOf(newFilter, filter);
  return newFilter;
};
