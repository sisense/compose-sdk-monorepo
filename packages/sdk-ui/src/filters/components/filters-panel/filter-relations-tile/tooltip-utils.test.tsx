import { describe, it, expect } from 'vitest';
import { generateTooltipLines, trimUnnecessaryBrackets } from './tooltip-utils';
import { TFunction } from '@ethings-os/sdk-common';
import { ReactNode } from 'react';
import { FilterRelationsDescription } from '@/utils/filter-relations';

describe('tooltip-utils', () => {
  describe('generateTooltipLines', () => {
    const mockT = ((key: string) => {
      const translations: Record<string, string> = {
        'filterRelations.and': 'AND',
        'filterRelations.or': 'OR',
      };
      return translations[key] || key;
    }) as unknown as TFunction;

    it('should generate tooltip with attributes and operators', () => {
      const filterRelationsDescription: FilterRelationsDescription = [
        { nodeType: 'openBracket' },
        { nodeType: 'attribute', attribute: 'Age' },
        { nodeType: 'operator', operator: 'AND' },
        { nodeType: 'attribute', attribute: 'Salary' },
        { nodeType: 'closeBracket' },
      ];

      const result: ReactNode[] = generateTooltipLines(filterRelationsDescription, mockT);

      expect(result).toMatchSnapshot();
    });

    it('should handle multiple operators and brackets', () => {
      const filterRelationsDescription: FilterRelationsDescription = [
        { nodeType: 'openBracket' },
        { nodeType: 'attribute', attribute: 'Age' },
        { nodeType: 'operator', operator: 'OR' },
        { nodeType: 'attribute', attribute: 'Salary' },
        { nodeType: 'closeBracket' },
        { nodeType: 'operator', operator: 'AND' },
        { nodeType: 'attribute', attribute: 'Location' },
      ];

      const result: ReactNode[] = generateTooltipLines(filterRelationsDescription, mockT);

      expect(result).toMatchSnapshot();
    });

    it('should handle empty description', () => {
      const filterRelationsDescription: FilterRelationsDescription = [];

      const result: ReactNode[] = generateTooltipLines(filterRelationsDescription, mockT);

      expect(result).toMatchObject([<></>]);
    });
  });

  describe('trimUnnecessaryBrackets', () => {
    it('should remove outermost unnecessary brackets', () => {
      const filterRelationsDescription: FilterRelationsDescription = [
        { nodeType: 'openBracket' },
        { nodeType: 'attribute', attribute: 'Age' },
        { nodeType: 'operator', operator: 'AND' },
        { nodeType: 'attribute', attribute: 'Salary' },
        { nodeType: 'closeBracket' },
      ];

      const result = trimUnnecessaryBrackets(filterRelationsDescription);

      expect(result).toEqual([
        { nodeType: 'attribute', attribute: 'Age' },
        { nodeType: 'operator', operator: 'AND' },
        { nodeType: 'attribute', attribute: 'Salary' },
      ]);
    });

    it('should return the original description if fewer than two nodes', () => {
      const filterRelationsDescription: FilterRelationsDescription = [
        { nodeType: 'attribute', attribute: 'Age' },
      ];

      const result = trimUnnecessaryBrackets(filterRelationsDescription);

      expect(result).toEqual(filterRelationsDescription);
    });

    it('should handle deeply nested brackets', () => {
      const filterRelationsDescription: FilterRelationsDescription = [
        { nodeType: 'openBracket' },
        { nodeType: 'openBracket' },
        { nodeType: 'attribute', attribute: 'Age' },
        { nodeType: 'closeBracket' },
        { nodeType: 'closeBracket' },
      ];

      const result = trimUnnecessaryBrackets(filterRelationsDescription);

      expect(result).toEqual([{ nodeType: 'attribute', attribute: 'Age' }]);
    });
  });
});
