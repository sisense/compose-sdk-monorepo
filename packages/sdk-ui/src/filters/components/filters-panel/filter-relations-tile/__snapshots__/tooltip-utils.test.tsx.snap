// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tooltip-utils > generateTooltipLines > should generate tooltip with attributes and operators 1`] = `
[
  <React.Fragment>
    <i>
      Age
    </i>
  </React.Fragment>,
  <b>
    AND
  </b>,
  <React.Fragment>
    <i>
      Salary
    </i>
  </React.Fragment>,
]
`;

exports[`tooltip-utils > generateTooltipLines > should handle multiple operators and brackets 1`] = `
[
  <React.Fragment>
    <React.Fragment>
      (
    </React.Fragment>
    <i>
      Age
    </i>
  </React.Fragment>,
  <b>
    <b>
      OR
    </b>
  </b>,
  <React.Fragment>
    <i>
      Salary
    </i>
    <React.Fragment>
      )
    </React.Fragment>
  </React.Fragment>,
  <b>
    AND
  </b>,
  <React.Fragment>
    <i>
      Location
    </i>
  </React.Fragment>,
]
`;
